servicer/builtin/service_adapters/package/base_package.py

```diff
@@ -149,8 +156,20 @@ def read_package_info(self, package_info={}):
 
         if 'name' not in self.package_info:
             self.package_info['name'] = self.package_name(self.config['package_info']['package_file_path'])
+
+        try:
+            self.package_version_format
+        except AttributeError:
+            if 'package_version_format' in self.package_info:
+                self.package_version_format = self.package_info['package_version_format']
+            else:
+                raise NameError('Service must provide package_version_format.')
```

Comment:
If `package_version_format` is provided as part of `package_info`, it should override the initial value of `self.package_version_format` that comes from either the pypi or sbt adapter. Currently you can only use the default value if you're using the pypi or sbt adapters.

I think this try/except can be avoided by providing a default value for `self.package_version_format` in the base_package adapter's init method.
```
self.package_version_format = None
# or
self.package_version_format = '%s'
```

Then this block can be reduced to:
```
if 'package_version_format' in self.package_info:
    self.package_version_format = self.package_info['package_version_format']

if not self.package_version_format:
    raise ValueError('Service must provide package_version_format.') # NameError is for nonexistent local/global python variables, so ValueError makes a bit more sense here.
```

An adapter inheriting from base_package (or base_package itself) can then force the user to provide `package_version_format` in package_info by setting `self.package_version_format` to `None` in its init method.

Reply:
I'd originally provided defaults for the adapters. However, prohibiting services from defining package_version_format in package_info also prohibits them from using automatic versioning if they aren't using an adapter. For example, if you've got a project that saves its version in a setup.py but is delivered only as a Docker image, it'll need a base_package service that provides GCR as its existing_versions_source and a different package_version_format.

Reply:
This change shouldn't prevent a service using the base package adapter from providing `package_version_format`. If you initialize `self.package_version_format` to `None`, you'll be requiring that a service using the base package must provide `package_version_format`, which I think is your goal.

Reply:
Yup, that's cleaner.